### PR TITLE
Don't leak non-overwritten map settings to config file

### LINKED ISSUE #11351: Persistence of water_level=-31000
`minetest.conf` is the correct way for games to temporarily set settings so if water_level indeed persists beyond that this is an engine bug.

_Originally posted by @sfan5 in https://github.com/minetest/minetest/issues/11194#issuecomment-819862349_

> `minetest.conf` is the correct way for games to temporarily set settings so if water_level indeed persists beyond that this is an engine bug.

I'm not sure if I should necrobump this or not as I'm new to GitHub.
But anyway, this bug is reproducible for me.

Minetest Version: 5.4.1
- Tutorial by Wuzzy indeed set a global water_level of -31000 in minetest.conf and persists there.

### PR DESCRIPTION
**Pull request by sfan5**
It turns out that one global hierarchy cannot cover the needs of the map settings manager. Fortunately there's a simple workaround.

fixes #11351

## To do

This PR is Ready for Review.

## How to test

1) Have a mod that runs `minetest.set_mapgen_setting("water_level", "-31000")`
2) It should not appear in minetest.conf

### CODE REVIEW

**Comment by SmallJoker** [comment]
from what I can see, it would also be possible to add in another layer between `SL_GLOBAL` and `SL_MAP` named `SL_MAP_DEFAULTS` which is then written to in this special case. It would also keep the structured fallback order which I intended initially when rewriting the Settings hierarchy.

However, I do see that the map settings are not needed outside of `MapSettingsManager`. What I would suggest (without any new class):
1. new private field `Settings::m_parent`, to be used in `MapSettingsManager`. points to a "map defaults" layer.
2. `*m_parent` ("map defaiults") would then itself have a parent to the global hierarchy entry point (i.e. `SL_GLOBAL`)
3. `Settings::getParent()` would then either check `m_parent` if specified, or, if `this` is within the global hierarchy (i.e. `m_settingslayer` valid) iterate backwards.

**Comment by sfan5** [comment]
> From what I can see, it would also be possible to add in another layer between `SL_GLOBAL` and `SL_MAP` named `SL_MAP_DEFAULTS` which is then written to in this special case.

Sure that'd work, but for me a central problem is that map_meta should not be part of the global setting layers at all.
The global hierarchy should only deal with settings in the same "namespace" as minetest.conf. Mapgen settings are a special case which happen to fall back to minetest.conf but also not exactly (see "seed").

It's also a coincidence that putting it on top of `SL_GLOBAL` works.
As a theoretical example if we additionally had `Settings` for env_meta.txt and those should fall back to the global ones, that would be impossible with the current setup. Or a fallback chain that skips game settings for some reason, that wouldn't work either.

> It would also keep the structured fallback order which I intended initially when rewriting the Settings hierarchy.

Not sure what you mean, the fallback order is exactly the same.

> I would like to suggest an alternative to adding a "supervisor" class with callbacks: new private field `Settings::m_parent`

I think that would be a worse solution, `Setting`s objects are still forced to "participate" in the global layers except there would be a way to bypass that, implying that the abstraction doesn't work/is incomplete.
`SettingsHierarchy` is certainly the cleaner solution here since both the global and any custom ones are implemented using the exact same abstraction without one having the unfair advantage of being hardcoded inside `Settings`.

**Comment by SmallJoker** [comment]
Discussion reference: https://irc.minetest.net/minetest-dev/2021-06-16#i_5832945

TL;DR: A "supervisor" class is necessary to properly deal with destroyed `Settings` during runtime.

File: src/main.cpp [hunk 1]
```diff
@@ -490,7 +493,6 @@ static bool init_common(const Settings &cmd_args, int argc, char *argv[])
 
 	// Initialize sockets
 	sockets_init();
-	atexit(sockets_cleanup);
```
Comment (nerzhul) [inline src/main.cpp:493]: why not call uninit_common here ?

File: src/main.cpp [hunk 2]
```diff
@@ -513,6 +515,17 @@ static bool init_common(const Settings &cmd_args, int argc, char *argv[])
 	return true;
 }
 
+static void uninit_common()
```
Comment (nerzhul) [inline src/main.cpp:518]: instead of uninin, which is crazy to read, on_shutdown_common is better :)

File: src/map_settings_manager.cpp [hunk 3]
```diff
@@ -26,15 +26,24 @@ with this program; if not, write to the Free Software Foundation, Inc.,
...
-	m_map_settings = Settings::createLayer(SL_MAP, "[end_of_params]");
-	Mapgen::setDefaultSettings(Settings::getLayer(SL_DEFAULTS));
+	/*
+	 * We build our own hierarchy which falls back to the global one.
+	 * It looks as follows: (lowest prio first)
+	 * 0: whatever is picked up from g_settings (incl. engine defaults)
+	 * 1: defaults set by scripts (override_meta = false)
+	 * 2: settings present in map_meta.txt or overriden by scripts
+	 */
+	m_defaults = new Settings("", &m_hierarchy, 1);
```
Comment (nerzhul) [inline src/map_settings_manager.cpp:39]: Why use a pointer here ? Cannot we just put Settings on the stack instead ?

File: src/settings.cpp [hunk 4]
```diff
@@ -33,48 +33,100 @@ with this program; if not, write to the Free Software Foundation, Inc.,
...
 
-Settings *Settings::s_layers[SL_TOTAL_COUNT] = {0}; // Zeroed by compiler
 std::unordered_map<std::string, const FlagDesc *> Settings::s_flags;
 
+/* Settings hierarchy implementation */
 
-Settings *Settings::createLayer(SettingsLayer sl, const std::string &end_tag)
+SettingsHierarchy::SettingsHierarchy(Settings *fallback)
+{
+	layers.resize(1);
```
Comment (nerzhul) [inline src/settings.cpp:46]: no nee to resize + set, there is a single element. Just push_back

File: src/settings.cpp [hunk 5]
```diff
@@ -33,48 +33,100 @@ with this program; if not, write to the Free Software Foundation, Inc.,
...
+	for (int i = layer - 1; i >= 0; --i) {
+		if (layers[i])
+			return layers[i];
+	}
+
+	return nullptr;
+}
+
+
+void SettingsHierarchy::layerCreated(int layer, Settings *obj)
```
Comment (nerzhul) [inline src/settings.cpp:72]: layer should be a unsigned int

File: src/settings.cpp [hunk 6]
```diff
@@ -33,48 +33,100 @@ with this program; if not, write to the Free Software Foundation, Inc.,
...
+		throw BaseException("Setting layer " + itos(layer) + " already exists");
+
+	pos = obj;
+	// This feels bad
+	if (this == &g_hierarchy && layer == (int)SL_GLOBAL)
+		g_settings = obj;
+}
+
+
+void SettingsHierarchy::layerRemoved(int layer)
```
Comment (nerzhul) [inline src/settings.cpp:90]: u32

File: src/settings.h [hunk 7]
```diff
@@ -214,6 +239,10 @@ class Settings {
 
 	void removeSecureSettings();
 
+	// Returns the settings layer this object is.
+	// If within the global hierarchy you can cast this to enum SettingsLayer
+	inline int getLayer() const { return m_settingslayer; }
```
Comment (nerzhul) [inline src/settings.h:244]: u32

File: src/settings.h [hunk 8]
```diff
@@ -257,7 +286,8 @@ class Settings {
 	// All methods that access m_settings/m_defaults directly should lock this.
 	mutable std::mutex m_mutex;
 
-	static Settings *s_layers[SL_TOTAL_COUNT];
-	SettingsLayer m_settingslayer = SL_TOTAL_COUNT;
+	SettingsHierarchy *m_hierarchy = nullptr;
+	int m_settingslayer = -1;
```
Comment (nerzhul) [inline src/settings.h:290]: u32

**Review by nerzhul — CHANGES_REQUESTED** [review]

File: src/map_settings_manager.cpp (hunk 3 above)
Comment (sfan5) [reply src/map_settings_manager.cpp:39]: Probably but
* The initialization order and destruction order needs to be gotten right since the settings objects depend on `m_hierarchy`
* I would have to change existing code to use `.` instead of `->`

File: src/main.cpp (hunk 1 above)
Comment (sfan5) [reply src/main.cpp:493]: Using `atexit` you mean? I think it's cleaner when we call it manually.

File: src/settings.h (hunk 8 above)
Comment (sfan5) [reply src/settings.h:290]: How would I set this to -1 if it's unsigned? (That is supposed to represent an invalid layer number)

File: src/main.cpp (hunk 1 above)
Comment (SmallJoker) [reply src/main.cpp:493]: I support nerzhul's request. `atexit` is called reliably on sane program termination, hence should IMO be used for this cleanup.

File: src/settings.cpp (hunk 5 above)
Comment (SmallJoker) [reply src/settings.cpp:72]: Yes, `-1` layer values are not really necessary. `Settings::m_hierarchy` can already be used to determine whether it's part of a structure, or just any loose Settings instance in the wild.

File: src/settings.cpp (hunk 6 above)
Comment (SmallJoker) [reply src/settings.cpp:90]: `onLayerRemove`. The instance is yet not deleted.

File: src/settings.cpp (hunk 5 above)
Comment (sfan5) [reply src/settings.cpp:72]: It does not create the settings object itself so I think `onLayerCreated` is better.

File: src/settings.h (hunk 8 above)
Comment (nerzhul) [reply src/settings.h:290]: @sfan5 is this normal to have a Setting without layer ? i think we can have one layer by default, else this has no sense. Do you have usecase without layers ?

Comment (nerzhul) [reply src/settings.h:290]: For me layer 0 is the root.

```
setting name: setting.aaa.bbb.ccc
setting layer:      0   1   2   3
```

Comment (sfan5) [reply src/settings.h:290]: Setting objects can exist without being bound to any hierarchy, then it makes no sense to define any layer.
I don't know what the second post is supposed to mean.

**Review by SmallJoker — APPROVED** [review]
works.
